hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/table/HoodieTableSink.java

```diff
@@ -103,6 +103,8 @@ public SinkRuntimeProvider getSinkRuntimeProvider(Context context) {
           conf.setBoolean(FlinkOptions.COMPACTION_ASYNC_ENABLED, false);
         }
         return Pipelines.compact(conf, pipeline);
+      } else if (OptionsResolver.isMorTable(conf)) {
+        return Pipelines.dummySink(pipeline);
       } else {
```

Comment:
So what component is responsible for data cleaning then?

Reply:
Offline compaction do clean，Consistent with clustering.
  
    // Append mode
      if (OptionsResolver.isAppendMode(conf)) {
        DataStream<Object> pipeline = Pipelines.append(conf, rowType, dataStream, context.isBounded());
        if (OptionsResolver.needsAsyncClustering(conf)) {
          return Pipelines.cluster(conf, rowType, pipeline);
        } else {
          return Pipelines.dummySink(pipeline);// If execute-clustering=true, no clean.
        }
      }

Reply:
The cleaning can take effect finally right? Because the table would get compacted for sometime anyway.

Reply:
Offline clustering/compaction will do clean by default ,   add --clean-async-enabled config will close it.

Reply:
Add a sync-clean config verification for this.

Reply:
Current online async-clean and offline clean use the same configuration ：clean.async.enabled. Add a new configuration clean.offline.enable，making the configuration description clearer.

Reply:
Adjust the cleaning operation in SparkRDDWriteClient#cluster/compact,  when ASYNC_CLEAN is true will do clean in prewrite， otherwise will do clean in autoCleanOnCommit().

Reply:
We need to think through the e2e use case for flink streaming ingestion and spark offline compaction, we should add cleaning for spark compaction and clustering first which is a block change of this patch.

Reply:
We will not schedule cleaning task if async cleaning is disabled: https://github.com/apache/hudi/blob/67ae0c8e7e4e58454cce18a8f58bfa43f67c1183/hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/sink/CleanFunction.java#L72, so should be fine?

Reply:
If  CLEAN_ASYNC_ENABLED = true，a schedule will still be executed. I think cluster and compact should be consistent here. And now both Spark and Flink will clean up when executing offline jobs, unless forcibly closed. What do you think?